### PR TITLE
RELAETD: RAIL-4003 validElements for elements by value fix

### PR DESCRIPTION
**Pull request by plavemar**
Refactoring of the AttributeFilterButton component to improve the readability and maintainability of the component. 

- move cancelable promises to custom hooks
- extract 'inline' components to own files
- use backend property with telemetry only
- handle state in the custom hook only
- create utils specific for the AttributeFilterButton component

prerequisite to bugfixes RAIL-4003 and RAIL-4005 related to attribute elements defined by value.

JIRA: RAIL-4003, RAIL-4005

---

Supported PR commands:

| Command                  | Description            |
| ------------------------ | ---------------------- |
| `ok to test`             | Re-run standard checks |
| `extended test`          | BackstopJS tests       |
| `extended check sonar`   | SonarQube tests        |
| `extended check cypress` | Cypress E2E tests      |

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `check-extended` passes
-   [ ] `check-extended-cypress` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/6639/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/8527/ : SUCCESS in 10m 41s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/6683/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/8583/ : FAILURE in 3m 48s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/6692/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/8593/ : FAILURE in 3m 43s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/6695/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/8598/ : FAILURE in 3m 42s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/6700/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/8603/ : SUCCESS in 11m 36s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/6707/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/8610/ : SUCCESS in 11m 19s

File: libs/sdk-ui-filters/src/AttributeFilter/AttributeFilterButton/AttributeFilterButtonUtils.ts [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+            preparedElementQuery.withAttributeFilters(
+                getValidElementsFilters(parentFilters, parentFilterOverAttribute),
+            );
+        }
+    }
+
+    return preparedElementQuery;
+};
+
+export const prepareElementsTitleQuery = (
```
Comment (kandl) [inline libs/sdk-ui-filters/src/AttributeFilter/AttributeFilterButton/AttributeFilterButtonUtils.ts:97]: Should we put this as a local function to `useFetchInitialElements.ts`? (it's not used anywhere else)

File: libs/sdk-ui-filters/src/AttributeFilter/AttributeFilterButton/AttributeFilterButtonUtils.ts [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        isCancelablePromiseLoading(cancelablePromiseStatus) ||
+        isCancelablePromisePending(cancelablePromiseStatus)
+    );
+};
+
+export const getBackend = (backend: IAnalyticalBackend, props: any) => {
+    return backend.withTelemetry("AttributeFilter", props);
+};
+
+export const prepareElementsQuery = (
```
Comment (kandl) [inline libs/sdk-ui-filters/src/AttributeFilter/AttributeFilterButton/AttributeFilterButtonUtils.ts:62]: Should we put this as a local function to `useLoadMissingData.ts`? (it's not used anywhere else)

**Review by kandl — COMMENTED** [review]
Another useful thing to do (if you have it somehow in you head now) would be to comment purpose of some functions and props to make the upcoming refactor/rewrite easier.

File: libs/sdk-ui-filters/src/AttributeFilter/AttributeFilterButton/AttributeFilterButtonUtils.ts (hunk 1 above)
Comment (plavemar) [reply libs/sdk-ui-filters/src/AttributeFilter/AttributeFilterButton/AttributeFilterButtonUtils.ts:97]: The function moved to the hook

File: libs/sdk-ui-filters/src/AttributeFilter/AttributeFilterButton.tsx [hunk 3]
```diff
@@ -991,12 +412,44 @@ export const AttributeFilterButtonCore: React.FC<IAttributeFilterButtonProps> =
...
+            isFiltering={state.isFiltering}
+            isDropdownOpen={state.isDropdownOpen}
+            isElementsLoading={!state.validOptions?.items && isCancelablePromiseLoading(elementsStatus)}
+            isOriginalTotalCountLoading={isCancelablePromisePendingOrLoading(originalTotalCountStatus)}
+            title={
+                props.title || !isCancelablePromisePendingOrLoading(attributeStatus)
+                    ? attribute.title
+                    : getLoadingTitleIntl(props.intl)
+            }
+            subtitle={getSubtitle(
```
[the remark this answers is not in the document]
Comment (plavemar) [reply libs/sdk-ui-filters/src/AttributeFilter/AttributeFilterButton.tsx:425]: Parameters are structuralized to logical units.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

File: libs/sdk-ui-filters/src/AttributeFilter/AttributeFilterButton.tsx [hunk 4]
```diff
@@ -784,192 +349,48 @@ export const AttributeFilterButtonCore: React.FC<IAttributeFilterButtonProps> =
...
+        items: state.validOptions?.items ?? [],
+        totalCount: totalCount ?? ATTRIBUTE_FILTER_BUTTON_LIMIT,
+        onSelect: onElementSelect,
+        onRangeChange,
+        onSearch,
+        selectedItems: state.selectedFilterOptions,
+        isInverted: state.isInverted,
+        isLoading:
+            (!state.validOptions?.items && isCancelablePromiseLoading(elementsStatus)) ||
+            isCancelablePromisePendingOrLoading(totalCountStatus) ||
```
[the remark this answers is not in the document]
Comment (plavemar) [reply libs/sdk-ui-filters/src/AttributeFilter/AttributeFilterButton.tsx]: Good idea, the elementsStatus cannot be checked by this one-liner as we are checking only for "loading" status, not "loading" or "pending". Other statuses are checked this way.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/6710/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/8613/ : SUCCESS in 12m 43s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/6711/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/8614/ : SUCCESS in 11m 35s

**Comment by plavemar** [comment]
extended test

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/6733/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/1328/ : SUCCESS in 13m 40s

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR2214

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/1987/ : SUCCESS in 1m 06s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/2270/ : SUCCESS in 19s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/1078/ : SUCCESS in 15m 26s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/8636/ : SUCCESS in 11m 22s

**Pull request merged**